Fix RefreshIndicator sensitivity with NestedScrollView by using render size.

moritasoshi opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

This PR fixes overly sensitive pull-to-refresh behavior in RefreshIndicator when used with NestedScrollView that has large headers.

### Problem

When RefreshIndicator wraps a NestedScrollView with a large header (e.g., 550px), the inner scrollable viewport becomes very small (e.g., 50px). The refresh trigger threshold was previously calculated based on this small viewport dimension (50px × 0.25 = 12.5px), making the indicator extremely sensitive—even a slight 13px drag would trigger refresh. This eliminates the visual drag feedback users expect.

### Solution

Changed the threshold calculation to always use the RefreshIndicator's actual rendered size instead of the viewport dimension reported by the surrounding Scrollable. This provides more natural drag sensitivity for all layouts, especially those with small scrollable areas.

With this change, the threshold is based on the full RefreshIndicator size (e.g., 600px × 0.25 = 150px), requiring a more reasonable drag distance to trigger refresh.

### Changes

- Modified `_checkDragOffset()` to use `RenderBox.size.height` of the RefreshIndicator itself instead of viewport dimensions
- Added comprehensive test demonstrating the fix for NestedScrollView with large headers

Fixes #147363

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

moritasoshi committed:
Add an option to RefreshIndicator to use the actual viewport dimensions.

moritasoshi committed:
Merge branch 'master' into add-option-to-refreshindicator-to-use-actual-viewport

moritasoshi committed:
Rename useActualViewportDimension to useIndicatorExtent in RefreshIndicator

moritasoshi committed:
Remove the private method reference from public API docs

victorsanni commented on packages/flutter/lib/src/material/refresh_indicator.dart:
```diff
@@ -528,6 +546,19 @@ class RefreshIndicatorState extends State<RefreshIndicator>
+
+    final RenderObject? renderObject = context.findRenderObject();
+    if (renderObject is RenderBox && renderObject.hasSize) {
+      return renderObject.size.height;
```
I'm wondering if this is one of those bugs that only happens if containerExtent > size or vice versa (in that case, we can just use the max/min). Can you explain more here?

victorsanni reviewed (commented):
It would be nice if we can fix this bug without increasing the API surface.

moritasoshi commented on packages/flutter/lib/src/material/refresh_indicator.dart:
```diff
@@ -528,6 +546,19 @@ class RefreshIndicatorState extends State<RefreshIndicator>
     }
   }
 
+  double _resolveContainerExtent(double containerExtent) {
```
Yeah, thanks.

I was a bit hesitant because it would change the behavior slightly.

But it seems like any reasonable code wouldn't experience regressions. I'll modify the code to always use the size from the context.

moritasoshi replied:
Sure! I’ve removed the flag and updated it to always use the size.
The fix is complete, so please go ahead and run the Google testing CI check.

moritasoshi committed:
Use render size for RefreshIndicator threshold

victorsanni committed:
Merge branch 'master' into add-option-to-refreshindicator-to-use-actual-viewport

moritasoshi committed:
Merge branch 'master' into add-option-to-refreshindicator-to-use-actual-viewport

moritasoshi committed:
Remove obvious type annotations to fix lint warnings

moritasoshi committed:
Merge remote-tracking branch 'upstream/master' into add-option-to-refreshindicator-to-use-actual-viewport

victorsanni commented on packages/flutter/test/material/refresh_indicator_test.dart:
```diff
@@ -1270,4 +1270,72 @@ void main() {
+    await tester.pump(const Duration(seconds: 1));
+    expect(refreshCalled, false);
+
+    // A drag meeting the render-size-based threshold should trigger refresh
```
```suggestion
    // A drag meeting the render-size-based threshold should trigger refresh.
```

victorsanni commented on packages/flutter/test/material/refresh_indicator_test.dart:
```diff
@@ -1270,4 +1270,72 @@ void main() {
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
```
here also.

MitchellGoodwin reviewed (approved):
LGTM as well pending Victor's suggestions.

moritasoshi committed:
Add missing period to test comment

moritasoshi committed:
Refactor RefreshIndicator test to use pumpAndSettle

moritasoshi commented on packages/flutter/test/material/refresh_indicator_test.dart:
```diff
@@ -1270,4 +1270,72 @@ void main() {
+    await tester.pump();
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
+    await tester.pump(const Duration(seconds: 1));
```
Thank you for your comment.
I've updated the test to use `pumpAndSettle()` instead. The multiple pumps were needed because `onRefresh` is only called after the RefreshIndicator's snap animation finishes.

Piinks commented on packages/flutter/lib/src/material/refresh_indicator.dart:
```diff
@@ -514,8 +514,10 @@ class RefreshIndicatorState extends State<RefreshIndicator>
-  void _checkDragOffset(double containerExtent) {
+  void _checkDragOffset() {
     assert(_status == RefreshIndicatorStatus.drag || _status == RefreshIndicatorStatus.armed);
+    final renderBox = context.findRenderObject()! as RenderBox;
```
I think so. This is a clever solution since the NestedScrollView's scroll notification metrics represent the individual scroll views rather than the sum of both. The refresh indicator wraps the whole Nested scroll view, so it has an accurate dimension for the full nested scrolling window.


We should put some assertions in place though. 
I would consider removing the `!` or adding a clear error message - for example if the context is invalidated while the gesture is being handled here.
Also, you should check that renderBox.hasSize before accessing size, if not that would also crash.

Piinks reviewed (changes requested):
Thanks for fixing this! Just a few adjustment to make before we can land it. There have been other PRs that tried to fix this, it looks like this is the right approach. 👍

Piinks reviewed (commented):
For the failing google test here, I would also check and consider how this change impacts this type of composition:

Column -> Expanded -> RefreshIndicator -> Container -> SingleChildScrollView (always scrollable scroll physics) -> with minimal content, not enough to scroll. Currently the refresh indicator can be used in this case, but with this change it breaks.

I can see about forming a runnable test case, I need to extract the Google bits.

moritasoshi committed:
Add null safety check

moritasoshi commented:
@victorsanni 
Thanks for the detailed explanation. I understand the issue now.
The core problem is that these two use cases require different reference values for calculating the drag threshold:

1. NestedScrollView with large headers: The containerExtent (from ScrollNotification) represents only the inner scrollable viewport, which can be much smaller than the visible area the user is interacting with.
2. RefreshIndicator wrapping padded content: The RefreshIndicator's rendered size includes non-scrolling areas (like padding), which is larger than the actual scrollable viewport.

In case 1, containerExtent is too small. In case 2, the rendered size is too large. There's no single value that correctly represents the "intended scrollable area" for both cases.
Given this fundamental conflict, I believe an optional parameter is necessary to let developers specify which calculation is appropriate for their layout.

Piinks commented:
Thank you for your contribution. I'm going to close this PR for now since there are outstanding comments, just to get this off our PR review queue. Please don't hesitate to re-open or submit a new PR if you have the time to address the review comments. Thanks!

moritasoshi commented:
Hi @Piinks, I've reworked the approach based on your feedback. Could you reopen this PR?

The fix now **only targets the NestedScrollView case** instead of impacting all scrollables:

```dart
if (notification.depth > 0) {
  final NestedScrollViewState? nsvState =
      notification.context?.findAncestorStateOfType<NestedScrollViewState>();
  if (nsvState != null) {
    final RenderBox? nsvRenderBox = nsvState.context.findRenderObject() as RenderBox?;
    if (nsvRenderBox != null && nsvRenderBox.hasSize) {
      containerExtent = nsvRenderBox.size.height;
    }
  }
}
```

- `depth > 0` guard ensures depth==0 scrollables (including padded SingleChildScrollView) are unaffected
- `findAncestorStateOfType<NestedScrollViewState>()` ensures only NestedScrollView cases use the adjusted extent
- Null safety is fully guarded with fallback to original `viewportDimension`
- Added regression tests for both NestedScrollView and non-NestedScrollView nested cases

Piinks commented:
Thanks for coming back! I am not sure isolating to NestedScrollView is the right approach either. Our style guide calls out that widgets should not have knowledge of each other, or behave differently based on the widget tree structure. This can lead to a poor developer experience when a widget behaves one way in one context, and quite differently in another.
I am also not sure, does this resolve the issue in miscalculation if the scroll view in the body of the nested scroll view in surrounded by other widgets like Padding?
Let's move discussion into the issue to see if we can work out a solution. :)
